Updating SIW OV2OV setup instructions

vakaevich-okta opened the pull request:
## Description:

Updating SIW OV2OV setup instructions

## PR Checklist

- [ ] Have you verified the basic functionality for this change?
- [ ] Did you add tests, as appropriate, following our [Automated Test guidelines](https://oktawiki.atlassian.net/wiki/spaces/eng/pages/2676497890/Automated+Testing+in+the+Signin+Widget)?
- [ ] Did you follow our [Security Best Practices](https://oktawiki.atlassian.net/wiki/display/eng/Security+Best+practices)?
- [ ] Did you verify the change by running [downstream monolith artifact](https://oktawiki.atlassian.net/wiki/spaces/eng/pages/102897979/Sign-in+Widget+Development#Sign-inWidgetDevelopment-Instructionstocreateandrunthedownstreamartifact(d16t))? (Provide link to build below)
- [ ] Does this PR include noticeable changes to the UI? (If yes, attach screenshots/video below)

### Issue:

- [OKTA-642165](https://oktainc.atlassian.net/browse/OKTA-642165)

### Reviewers:
@yannongli-okta , @helenchen-okta , @SohamKolhatkar-okta 

### Screenshot/Video:

One and two devices:
<img width="1193" alt="Screenshot 2023-09-11 at 17 50 38" src="https://github.com/okta/okta-signin-widget/assets/1379394/ed00e72f-2b9a-4b11-a66c-2b2d079ee71b">

No devices:
<img width="1187" alt="Screenshot 2023-09-11 at 17 48 27" src="https://github.com/okta/okta-signin-widget/assets/1379394/4faaa941-0a17-43de-a3e0-29f00bf0dcc4">



### Downstream Monolith Build:

helenchen-okta reviewed (commented):
Please have screenshots showing with this change what the below scenarios look like:
- User has two+ OV enrolled
- User has one OV enrolled
- User has no OV enrolled

helenchen-okta commented on src/v2/view-builder/views/ov/EnrollChannelPollDescriptionView.js:
```diff
@@ -76,12 +76,19 @@ export default View.extend({
   getTemplateData() {
     const contextualData = this.options.appState.get('currentAuthenticator').contextualData;
+    let enrolledDeviceName = '';
+    if (contextualData && contextualData?.devicebootstrap && contextualData?.devicebootstrap.enrolledDevices) {
```
Not familiar with js, but when would `contextualData?.devicebootstrap.enrolledDevices` return true? And what is the type? Would it return true in case of empty string?
Asking because below code would return enrolledDevices in the raw form if it cannot be converted to an array.

yannongli-okta replied:
May i know why we specifically parse the devices name here instead of just using the old way? any issue you spot with the old way?

vakaevich-okta commented:
@helenchen-okta , @yannongli-okta I've added test cases for multiple and no devices:
https://github.com/okta/okta-signin-widget/pull/3362/commits/536f1719329cc4d3f97c01dff216d110a30dc625
and added screenshots.
Please notice the "no device names" screenshot particularly, and let me know if we should use some other string for such case

helenchen-okta commented on test/testcafe/spec/EnrollAuthenticatorOktaVerify_spec.js:
```diff
@@ -250,7 +255,8 @@ const sameDeviceInstruction1 = 'If you don’t have Okta Verify installed, downl
 const deviceBootstrapSubtitle = 'To set up Okta Verify on additional devices, you can copy an existing Okta Verify account onto a new device.';
-const deviceBootstrapInstruction1 = 'Open Okta Verify on any of your other Okta Verify devices';
+const deviceBootstrapInstruction1 = 'Open Okta Verify on any of your other Okta Verify devices (Such as your testDevice1).';
+const deviceBootstrapInstructionNoDeviceName = 'Open Okta Verify and follow the steps to add your account.';
```
Can remove this?

vakaevich-okta replied:
sure: https://github.com/okta/okta-signin-widget/pull/3362/commits/723ebb39c37cd6817a2bf9897472d06492f3beaa

yannongli-okta commented on test/testcafe/spec/EnrollAuthenticatorOktaVerify_spec.js:
```diff
@@ -860,13 +865,36 @@ test
     await t.expect(enrollOktaVerifyPage.getSubHeader()).eql(deviceBootstrapSubtitle);
-    await t.expect(await enrollOktaVerifyPage.getNthInstructionBulletPoint(0)).contains(deviceBootstrapInstruction1);
-    await t.expect(await enrollOktaVerifyPage.getNthInstructionBulletPoint(0)).contains('testDevice1');
+    await t.expect(await enrollOktaVerifyPage.getNthInstructionBulletPoint(0)).eql(deviceBootstrapInstruction1);
```
pls assert we don't have device2 name

vakaevich-okta replied:
sure, added the assertion: https://github.com/okta/okta-signin-widget/pull/3362/commits/cf146c07a86f4f40973071cbd1673aba00ab1ce0
though on this line we assert equality to `Open Okta Verify on any of your other Okta Verify devices (Such as your testDevice1).`

jaredperreault-okta commented on src/v2/view-builder/views/ov/EnrollChannelPollDescriptionView.js:
```diff
@@ -76,12 +76,20 @@ export default View.extend({
+    if (contextualData && contextualData?.devicebootstrap && contextualData?.devicebootstrap.enrolledDevices) {
+      const enrolledDevices = contextualData?.devicebootstrap.enrolledDevices;
+      enrolledDeviceName = Array.isArray(enrolledDevices) && !_.isEmpty(enrolledDevices) ?
+        enrolledDevices[0] : enrolledDevices;
```
I assume `enrolledDevices` needs to be a string, but based on this check

```
Array.isArray(enrolledDevices) && !_.isEmpty(enrolledDevices)
```

if `enrolledDevices` is an empty array, enrolledDevices would equal `[]`
Secondly, if `devicebootstrap.enrolledDevices` is a non-array, is it guaranteed to be a string?

vakaevich-okta replied:
`enrolledDevices` is expected to be an array of one string, but if there are more, we need to pick the first one. if just a string provided, that's also ok, eventually we need to pass to the `arguments` on line 62 either an array of one string, or just a string

jaredperreault-okta commented on packages/@okta/i18n/src/properties/login.properties:
```diff
@@ -1033,8 +1033,6 @@ oie.enroll.okta_verify.setup.email.title = Check your email
 oie.enroll.okta_verify.setup.sms.title = Check your text messages
 oie.enroll.okta_verify.setup.skipAuth.subtitle = To set up Okta Verify on additional devices, you can copy an existing Okta Verify account onto a new device.
 # {0} is a enrolled device's name
-oie.enroll.okta_verify.setup.skipAuth.openOv = Open Okta Verify on any of your other Okta Verify devices (<$1>{0}</$1>).
```
Was there a corresponding monolith change that removed this i18n key from monolith?

vakaevich-okta replied:
no, there is none

jaredperreault-okta replied:
Can you add this i18n key back in? This file needs to be in sync with that's in monolith. It's fine the string is no longer being used

vakaevich-okta replied:
Makes sense. Added: https://github.com/okta/okta-signin-widget/pull/3362/commits/ae6ecf74f07619f16b35c67cd41f91143cb6472a
